### DISCUSSION TITLE
Where to put NSPasteboard

### DISCUSSION THREAD

**Discussion by flexzuu** [progrium/darwinkit#20] [Ideas]
I was investigating NSPasteboard and got some api covered now

```go
package nspasteboard

import (
	"github.com/progrium/macdriver/core"
	"github.com/progrium/macdriver/objc"
)

type NSPasteboardType = core.NSString

var (
	NSPasteboardTypeString = NSPasteboardType(core.NSString_FromString("public.utf8-plain-text"))
)

type NSPasteboard struct {
	objc.Object
}

var NSPasteboard_ = NSPasteboard{objc.Get("NSPasteboard")}

func NSPasteboard_GeneralPasteboard() NSPasteboard {
	return NSPasteboard{NSPasteboard_.Get("generalPasteboard")}
}

func (pb NSPasteboard) ClearContents() {
	pb.Send("clearContents")
}

func (pb NSPasteboard) SetStringForType(s string, t NSPasteboardType) {
	pb.Send("setString:forType:", core.NSString_FromString(s), t)
}
func (pb NSPasteboard) StringForType(t NSPasteboardType) string {
	return pb.Send("stringForType:", t).String()
}
```

Its based on this https://developer.apple.com/documentation/appkit/nspasteboard?language=objc so appkit.
Shoud this be in a package `AppKit` then?

Also does anyone know how to handle globals as `core.NSString_FromString("public.utf8-plain-text")` is not nice. 

I needed to write a objc program to get that value 👎 
I am not that familiar with either objc or c-go so any help would be nice.
I also wrote some wrappers to make the api work with readers and writers but i am not sure yet if they provide any value
```go
func Paste(dst io.Writer) error {
	r := strings.NewReader(nspasteboard.NSPasteboard_General().StringForType(nspasteboard.NSPasteboardTypeString))
	io.Copy(dst, r)
	return nil
}

func Copy(src io.Reader) error {
	b := &strings.Builder{}
	io.Copy(b, src)

	nspasteboard.NSPasteboard_General().ClearContents()
	nspasteboard.NSPasteboard_General().SetStringForType(b.String(), nspasteboard.NSPasteboardTypeString)
	return nil
}
```

Let me know how we wanna proceed I saw there is some appkit stuff in the cocoa package so maybe we just put it there for now?

**Comment by TotallyGamerJet** [comment]
The Cocoa framework actually includes AppKit. "The most important Cocoa class libraries come packaged in two core frameworks for each platform: Foundation and AppKit for OS X, and Foundation and UIKit for iOS" [(src)](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CocoaFundamentals/WhatIsCocoa/WhatIsCocoa.html). Therefore, all appkit methods would go into the cocoa framework. Perhaps we should create a subpackage inside of Cocoa for Appkit things?

**Reply by flexzuu** [reply]
okay sounds like we wanna add this to package `cocoa` for now. We could think of having `AppKit` and  `UIKit` packages at some point. 
I might open a PR that adds this then tonight.

**Comment by flexzuu** [comment]
PR is open: #25

**Comment by progrium** [comment]
my 2 cents: fewer packages the better. also shallow over deep. since in objective-c there is no namespace at all, just having rough categories for organizational purposes is an improvement. until it makes sense otherwise, appkit and uikit can live in cocoa.
